File: lib/py/uapi/internal/binary/ClientBinaryEncoder.py
from typing import TYPE_CHECKING

from uapi.internal.binary.BinaryEncoder import BinaryEncoder
from uapi.internal.binary.BinaryEncoding import BinaryEncoding

if TYPE_CHECKING:
    from uapi.ClientBinaryStrategy import ClientBinaryStrategy


class ClientBinaryEncoder(BinaryEncoder):

    def __init__(self, binaryChecksumStrategy: 'ClientBinaryStrategy'):
        self.recentBinaryEncoders: dict[int, BinaryEncoding] = {}
        self.binaryChecksumStrategy = binaryChecksumStrategy

    def encode(self, message: list[object]) -> list[object]:
        from uapi.internal.binary.ClientBinaryEncode import client_binary_encode
        return client_binary_encode(message, self.recentBinaryEncoders, self.binaryChecksumStrategy)

    def decode(self, message: list[object]) -> list[object]:
        from uapi.internal.binary.ClientBinaryDecode import client_binary_decode
        return client_binary_decode(message, self.recentBinaryEncoders, self.binaryChecksumStrategy)

File: lib/ts/uapi/internal/binary/ClientBinaryEncoder.ts
import { ClientBinaryStrategy } from 'uapi/ClientBinaryStrategy';
import { clientBinaryEncode } from 'uapi/internal/binary/ClientBinaryEncode';
import { clientBinaryDecode } from 'uapi/internal/binary/ClientBinaryDecode';
import { BinaryEncoder } from './BinaryEncoder';
import { BinaryEncoding } from './BinaryEncoding';

export class ClientBinaryEncoder implements BinaryEncoder {
    private readonly recentBinaryEncoders: Map<number, BinaryEncoding>;
    private readonly binaryChecksumStrategy: ClientBinaryStrategy;

    constructor(binaryChecksumStrategy: ClientBinaryStrategy) {
        this.recentBinaryEncoders = new Map<number, BinaryEncoding>();
        this.binaryChecksumStrategy = binaryChecksumStrategy;
    }

    encode(message: any[]): any[] {
        return clientBinaryEncode(message, this.recentBinaryEncoders, this.binaryChecksumStrategy);
    }

    decode(message: any[]): any[] {
        return clientBinaryDecode(message, this.recentBinaryEncoders, this.binaryChecksumStrategy);
    }
}
